src/ifd.rs

```diff
@@ -0,0 +1,175 @@
+    mask_ifds: Option<Vec<MaskIFD>>,
+}
+
+impl ImageFileDirectories {
+    pub(crate) async fn open(cursor: &mut ObjectStoreCursor, ifd_offset: usize) -> Self {
+        let mut next_ifd_offset = Some(ifd_offset);
+
+        let mut image_ifds = vec![];
+        let mut mask_ifds = vec![];
+        while let Some(offset) = next_ifd_offset {
+            let ifd = ImageFileDirectory::read(cursor, offset).await;
+            next_ifd_offset = ifd.next_ifd_offset();
+            match ifd {
+                ImageFileDirectory::Image(image_ifd) => image_ifds.push(image_ifd),
+                ImageFileDirectory::Mask(mask_ifd) => mask_ifds.push(mask_ifd),
+            }
+        }
+
+        Self {
+            image_ifds,
+            // TODO: if empty, return None
+            mask_ifds: Some(mask_ifds),
+        }
+    }
+}
+
+fn value_as_usize(value: &Value) -> usize {
+    match value {
+        Value::Byte(v) => *v as usize,
+        Value::Short(v) => *v as usize,
+        Value::Signed(v) => *v as usize,
+        Value::SignedBig(v) => *v as usize,
+        Value::Unsigned(v) => *v as usize,
+        Value::UnsignedBig(v) => *v as usize,
+        _ => panic!("Not an integer"),
+    }
+}
+
+struct OptionalTags {
+    directory: Directory,
+}
+
+impl OptionalTags {
+    /// Check if the IFD contains a full resolution image (not an overview)
+    fn is_full_resolution(&self) -> bool {
+        if let Some(val) = self.directory.get(&Tag::NewSubfileType) {
+            // if self.NewSubfileType.value[0] == 0:
+            todo!()
+        } else {
+            true
+        }
+        // self.directory.contains_key(T)
+    }
+}
+
+/// An ImageFileDirectory representing Image content
+// TODO: required tags should be stored as rust-native types, not Value
+struct ImageIFD {
+    // Required tags
+    bits_per_sample: Value,
+    compression: Value,
+
+    /// The number of rows of pixels in the image.
+    /// Type = SHORT or LONG
+    image_height: u32,
+
+    /// The number of columns in the image, i.e., the number of pixels per row.
+    /// Type = SHORT or LONG
+    image_width: u32,
+
+    photometric_interpretation: Value,
+    planar_configuration: Value,
+    sample_format: Value,
+    samples_per_pixel: Value,
+    tile_byte_counts: Value,
+    tile_height: Value,
+    tile_offsets: Value,
+    tile_width: Value,
+
+    directory: Directory,
+    next_ifd_offset: Option<usize>,
+}
+
+impl ImageIFD {
+    fn image_height(&self) -> u32 {
+        match self.image_height {
+            Value::Short(val) => val as u32,
+            Value::Unsigned(val) => val,
+            _ => unreachable!(),
+        }
+    }
+
+    fn image_width(&self) -> u32 {
+        match self.image_width {
+            Value::Short(val) => val as u32,
+            Value::Unsigned(val) => val,
+            _ => unreachable!(),
+        }
+    }
+
+    fn bands(&self) -> usize {
+        value_as_usize(&self.samples_per_pixel)
+    }
```

Comment:
Honestly, this looks wayyyy too easy after what I had to do in https://github.com/image-rs/image-tiff/pull/224/files :laughing: